KnapsackUtilsCython.pyx

```diff
@@ -0,0 +1,216 @@
+                sub_max = 0
+                prev_group = groups[i - 1] - 1
+                subKRow = curCapacity - weights[i - 1]
+                if prev_group > -1:
+                    prevGroupStart, prevGroupEnd = groupStartEnds[prev_group]
+                    j = prevGroupStart + 1
+                    while j <= prevGroupEnd:
+                        if groups[j - 1] == prev_group and arrget(K, n+1, j, subKRow) > sub_max:
+                            sub_max = arrget(K, n+1, j, subKRow)
+                        j += 1
+                        
+                arrset(K, n+1, i, curCapacity, max(sub_max + values[i - 1], arrget(K, n+1, i - 1, curCapacity)))
+            else:
+                arrset(K, n+1, i, curCapacity, arrget(K, n+1, i - 1, curCapacity))
+            i += 1
+        curCapacity += 1
+    res = arrget(K, n+1, n, capacity)
+    timeTaken = time.perf_counter() - timeStart
+    if not noLog:
+        logbook.info(f"Value Found {res} in {timeTaken:.3f}")
+    includedItems = []
+    includedGroups = []
+    w = capacity
+    lastTakenGroup = -1
+    i = n
+    while i > 0:
+        if res <= 0:
+            break
+        if i == 0:
+            raise AssertionError(f"i == 0 in knapsack items determiner?? res {res} i {i} w {w}")
+        if w < 0:
+            raise AssertionError(f"w < 0 in knapsack items determiner?? res {res} i {i} w {w}")
+        # either the result comes from the
+        # top (K[i-1][w]) or from (val[i-1]
+        # + K[i-1] [w-wt[i-1]]) as in Knapsack
+        # table. If it comes from the latter
+        # one/ it means the item is included.
+        # THIS IS WHY VALUE MUST BE INTS
+        if res == arrget(K, n+1, i - 1, w):
+            i -= 1
+            continue
+
+        group = groups[i - 1]
+        if group == lastTakenGroup:
+            i -= 1
+            continue
+
+        includedGroups.append(group)
+        lastTakenGroup = group
+        # This item is included.
+        if not noLog:
+            logbook.info(
+                f"item at index {i - 1} with value {values[i - 1]} and weight {weights[i - 1]} was included... adding it to output. (Res {res})")
+        includedItems.append(items[i - 1])
+
+        # Since this weight is included
+        # its value is deducted
+        res = res - values[i - 1]
+        w = w - weights[i - 1]
+        
+        i -= 1
+
+    if not noLog:
+        uniqueGroupsIncluded = set(includedGroups)
+        if len(uniqueGroupsIncluded) != len(includedGroups):
+            raise AssertionError("Yo, the multiple choice knapsacker failed to be distinct by groups")
+        logbook.info(
+            f"multiple choice knapsack completed on {n} items for capacity {capacity} finding value {arrget(K, n+1, n, capacity)} in Duration {time.perf_counter() - timeStart:.3f}")
+
+    return arrget(K, n+1, n, capacity), includedItems
+
+
+# cdef cython.int arrget(np.ndarray arr, cython.int dim, cython.int i, cython.int j):
```

Comment:
numpy performed worse I'm guessing? My experience with numpy is that it is good for SIMD, not for direct access stuff, but maybe in C direct is better. Also I may be mixing Pandas with Numpy

Reply:
It was slower than plain MemoryView, but actually some sources claim that it performs okay if you disable bounds checks. I'll need to research it.